File: algorithms/leetcode/572-subtree-of-another-tree/SubtreeOfAnotherTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private boolean checkSubtree(TreeNode root, TreeNode t) {
        if (root == null && t == null) {
            return true;
        }
        
        if (root == null || t == null) {
            return false;
        }
        
        if (root.val == t.val) {
            return checkSubtree(root.left, t.left) && checkSubtree(root.right, t.right);
        }
        
        return false;
    }
    
    public boolean isSubtree(TreeNode s, TreeNode t) {
        if (s == null) {
            return false;
        }
        
        return checkSubtree(s, t) || isSubtree(s.left, t) || isSubtree(s.right, t);
    }
}

File: leetcode/572-subtree-of-another-tree/subtree-of-another-tree.js
const isSubtree = (s, t) => {
    const checkSubTree = (root, root2) => {
        if (!root && !root2) {
            return true;
        }

        if (!root || !root2) {
            return false;
        }

        if (root.val !== root2.val) {
            return false;
        }

        return checkSubTree(root.left, root2.left) && checkSubTree(root.right, root2.right);
    };

    const traverse = node => {
        if (!node) {
            return false;
        }

        if (node.val === t.val) {
            if (checkSubTree(node, t)) {
                return true;
            }
        }

        return traverse(node.left) || traverse(node.right);
    };

    return traverse(s);
};
